codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -358,18 +357,23 @@ public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, lon
      */
     @Override
     public ChannelFuture writeGoAway(ChannelHandlerContext ctx, int lastStreamId, long errorCode, ByteBuf debugData,
-            ChannelPromise promise) {
-        Http2Connection connection = connection();
-        if (connection.isGoAway()) {
+                                     ChannelPromise promise) {
+        try {
+            if (connection().goAwaySent() && connection().remote().lastKnownStream() < lastStreamId) {
+                throw connectionError(PROTOCOL_ERROR, "Last stream identifier must not increase between " +
+                                                      "sending multiple GOAWAY frames (was '%d', is '%d').",
+                                                      connection().remote().lastKnownStream(),
+                                                      lastStreamId);
+            }
+            connection().goAwaySent(lastStreamId);
```

Comment:
> although I suspect we'll be closing the connection for other reasons anyway.

I suspect the same, but I don't think it will hurt to force the issue....to make our assumptions apparent right at the scene of the crime.